examples/json.dx

```diff
@@ -0,0 +1,364 @@
+     Area
+     Bar
+     Circle
+     Line
+     Point
+     Rect
+     Rule
+     Square
+     Text
+     Tick
+     
+instance Show Mark
+         show = (\ x.
+              case x of
+                   Area -> "area"
+                   Bar -> "bar"
+                   Circle -> "circle"
+                   Line -> "line"
+                   Point -> "point"
+                   Rect -> "rect"
+                   Rule-> "rule"
+                   Square -> "square"
+                   Tick -> "tick")
+
+data Channel =
+     Y
+     X
+     Color
+     Tooltip
+     HREF
+     Row
+     Col
+     Size
+     
+instance Show Channel
+         show = (\ x.
+              case x of
+                   Y -> "y"
+                   X -> "x"
+                   Color -> "color"
+                   Tooltip -> "tooltip"
+                   HREF -> "href"
+                   Size -> "size"
+                   Row -> "row"
+                   Col -> "col")
+
+' Most things in VL can take in extra visual options.
+  To avoid specifying these, we will take in as
+  JSON.
+
+data Opts a =
+     WithOpts a Value
+
+def pure (x:a) : Opts a =
+    WithOpts x AsNone
+    
+def pureLs (x:a) : List (Opts a) =
+   AsList 1 [WithOpts x AsNone]
+
+def mergeOpts [ToJSON a, ToJSON b] (x : a) (y : b) : Value =
+     case toJSON x of 
+         (AsObject x') -> case toJSON y of
+                  (AsObject y') -> AsObject $ x' <> y'
+                  (AsNone) -> AsObject x'
+         (AsNone) -> toJSON y
+ 
+def VLDataFrame (row:Type) (col:Type) (value:Type): Type =
+    DataFrame row col value (Header & EncodingType) 
+
+
+def chart [ToJSON v, ToJSON o] (x: VLDataFrame n (Fin key) v)
+                     (mark: Opts Mark)
+                     (encs : (Fin key) => List (Opts Channel))
+                     (opts : o)
+             : Value =
+
+    -- Make the mark
+    (WithOpts mtype options) = mark
+    jmark = ("mark", mergeOpts options [("type", show mtype)])
+
+    -- Make the data
+    (AsDataFrame meta df) = x
+    finsize = Fin $ size n
+    jdf = toJSON $ castTable finsize $ for i. toJSON $ for k.
+                   ("field" <> (show $ ordinal k),
+                   toJSON df.i.k)
+    jdata = ("data", toJSON [("values", jdf)])
+
+    -- Make the encodings
+    jencodings = toJSON $ concat $ for i.
+        (AsList v encopts) = encs.i
+        AsList _ $ for c : (Fin v). 
+           (WithOpts channel encoptions) = encopts.c
+           (title, type):(String & EncodingType) = meta.i
+           (show channel,
+            mergeOpts encoptions 
+            [
+              ("field", "field" <> (show $ ordinal i)),
+              ("type", show type),
+              ("title", title)
+              ])           
+    jencode = ("encoding", jencodings)
+    mergeOpts opts [jdata, jmark, jencode]
+    
+
+def showVega (x: Value) : String =
+    "<iframe width=\"100%\" frameborder=\"0\" scrolling=\"no\" onload=\"resizeIframe(this)\" srcdoc='<html> 
```

Comment:
Can't this `resizeIframe` business be done with something like `style="height:100%"` on a bunch of the tags below?

Reply:
Apparently you can't do this :( Need to resize with js as far as I can tell. But the JS seems to work.